File: core/clientLaya/game/src/shine/util/OtherUtils.ts
namespace Shine
{
	/** 杂项方法 */
	export class OtherUtils
	{
		/** 数组相等比较 */
		public static dataArrEquals(arr1: BaseData[], arr2: BaseData[]): boolean 
		{
			if (arr1 == null || arr2 == null)
				return false;

			var len: number;
			if ((len = arr1.length) != arr2.length)
				return false;

			for (var i: number = 0; i < len; ++i) 
			{
				if (!arr1[i].dataEquals(arr2[i]))
					return false;
			}

			return true;
		}

		// /** 移除不存在的项，从字典1,参考字典2 */
		// public static removeNotExistFromDic1WithDic2<T1, T2>(dic1: SMap<number, T1>, dic2: SMap<number, T2>): void 
		// {
		// 	var keys: number[] = dic1.getKeys();
		// 	var fv: number = dic1.getFreeValue();
		// 	var k: number;

		// 	for (var i: number = keys.length - 1; i >= 0; --i) 
		// 	{
		// 		if ((k = keys[i]) != fv) 
		// 		{
		// 			if (!dic2.contains(k)) 
		// 			{
		// 				dic1.remove(k);
		// 				++i;
		// 			}
		// 		}
		// 	}
		// }

		// /** 在字典中找寻最小的Key */
		// public static findLeastOfDic<T>(dic: SMap<number, T>, compare: Comparison<T>): number 
		// {
		// 	var keys: number[] = dic.getKeys();
		// 	var values: T[] = dic.getValues();
		// 	var fv: number = dic.getFreeValue();
		// 	var k: number;
		// 	var v: T;

		// 	var temp: T = null;
		// 	var has: boolean = false;
		// 	var re: number = -1;

		// 	for (var i: number = keys.length - 1; i >= 0; --i) 
		// 	{
		// 		if ((k = keys[i]) != fv) 
		// 		{
		// 			v = values[i];

		// 			if (!has) 
		// 			{
		// 				has = true;
		// 				temp = v;
		// 				re = k;
		// 			}
		// 			else 
		// 			{
		// 				if (compare(temp, v) < 0) 
		// 				{
		// 					temp = v;
		// 					re = k;
		// 				}
		// 			}
		// 		}
		// 	}

		// 	return re;
		// }

		//**********重载**********
		// /** 在字典中找寻最小的 */
		// public static findLeastOfDic<T>(dic:numberObjectMap<T> ,compare:Comparison<T> ):number
		// {
		// 	var keys:number[] =dic.getKeys();
		// 	var values:T[] =dic.getValues();
		// 	var fv:number =dic.getFreeValue();
		// 	var k:number ;
		// 	var v:T ;

		// 	T temp=default(T);
		// 	boolean has=false;
		// 	number re=-1;

		// 	for(number i=keys.length-1;i>=0;--i)
		// 	{
		// 		if((k=keys[i])!=fv)
		// 		{
		// 			v=values[i];

		// 			if(!has)
		// 			{
		// 				has=true;
		// 				temp=v;
		// 				re=k;
		// 			}
		// 			else
		// 			{
		// 				if(compare(temp,v)<0)
		// 				{
		// 					temp=v;
		// 					re=k;
		// 				}
		// 			}
		// 		}
		// 	}

		// 	return re;
		// }

		// /** 将元素放入有尺寸限制的集合中 */
		// public static putObjInDicWithMax<T>(key:number,obj:T,dic:SMap<number,T>, max:number,compare:Comparison<T>): void 
		// {
		// 	if (max <= 0) 
		// 	{
		// 		dic.put(key, obj);
		// 	}
		// 	else {
		// 		if (dic.contains(key)) {
		// 			dic.put(key, obj);
		// 		}
		// 		else {
		// 			if (dic.size() >= max) 
		// 			{
		// 				var leastKey: number = this.findLeastOfDic(dic, compare);

		// 				dic.remove(leastKey);
		// 			}

		// 			dic.put(key, obj);
		// 		}
		// 	}
		// }

		//**********重载**********
		// /** 将元素放入有尺寸限制的集合中 */
		// public static void putObjInDicWithMax<T>(number key,T obj,numberObjectMap<T> dic,number max,Comparison<T> compare)
		// {
		// 	if(max<=0)
		// 	{
		// 		dic.put(key,obj);
		// 	}
		// 	else
		// 	{
		// 		if(dic.contains(key))
		// 		{
		// 			dic.put(key,obj);
		// 		}
		// 		else
		// 		{
		// 			if(dic.size()>=max)
		// 			{
		// 				number leastKey=findLeastOfDic(dic,compare);

		// 				dic.remove(leastKey);
		// 			}

		// 			dic.put(key,obj);
		// 		}
		// 	}
		// }
	}
}

File: core/client/game/src/shine/utils/OtherUtils.cs
using System;

namespace ShineEngine
{
	/** 杂项方法 */
	public class OtherUtils
	{
		/** 数组相等比较 */
		public static bool dataArrEquals(BaseData[] arr1,BaseData[] arr2)
		{
			if(arr1==null || arr2==null)
				return false;

			int len;
			if((len=arr1.Length)!=arr2.Length)
				return false;

			for(int i=0;i<len;i++)
			{
				if(!arr1[i].dataEquals(arr2[i]))
					return false;
			}

			return true;
		}

		/** 移除不存在的项，从字典1,参考字典2 */
		public static void removeNotExistFromDic1WithDic2<T1,T2>(IntObjectMap<T1> dic1,IntObjectMap<T2> dic2)
		{
			int[] keys=dic1.getKeys();
			int fv=dic1.getFreeValue();
			int k;

			for(int i=keys.Length-1;i>=0;--i)
			{
				if((k=keys[i])!=fv)
				{
					if(!dic2.contains(k))
					{
						dic1.remove(k);
						++i;
					}
				}
			}
		}

		/** 在字典中找寻最小的Key */
		public static int findLeastOfDic<T>(IntObjectMap<T> dic,Comparison<T> compare)
		{
			int[] keys=dic.getKeys();
			T[] values=dic.getValues();
			int fv=dic.getFreeValue();
			int k;
			T v;

			T temp=default(T);
			bool has=false;
			int re=-1;

			for(int i=keys.Length-1;i>=0;--i)
			{
				if((k=keys[i])!=fv)
				{
					v=values[i];

					if(!has)
					{
						has=true;
						temp=v;
						re=k;
					}
					else
					{
						if(compare(temp,v)<0)
						{
							temp=v;
							re=k;
						}
					}
				}
			}

			return re;
		}

		/** 在字典中找寻最小的 */
		public static long findLeastOfDic<T>(LongObjectMap<T> dic,Comparison<T> compare)
		{
			long[] keys=dic.getKeys();
			T[] values=dic.getValues();
			long fv=dic.getFreeValue();
			long k;
			T v;

			T temp=default(T);
			bool has=false;
			long re=-1;

			for(int i=keys.Length-1;i>=0;--i)
			{
				if((k=keys[i])!=fv)
				{
					v=values[i];

					if(!has)
					{
						has=true;
						temp=v;
						re=k;
					}
					else
					{
						if(compare(temp,v)<0)
						{
							temp=v;
							re=k;
						}
					}
				}
			}

			return re;
		}

		/** 将元素放入有尺寸限制的集合中 */
		public static void putObjInDicWithMax<T>(int key,T obj,IntObjectMap<T> dic,int max,Comparison<T> compare)
		{
			if(max<=0)
			{
				dic.put(key,obj);
			}
			else
			{
				if(dic.contains(key))
				{
					dic.put(key,obj);
				}
				else
				{
					if(dic.size()>=max)
					{
						int leastKey=findLeastOfDic(dic,compare);

						dic.remove(leastKey);
					}

					dic.put(key,obj);
				}
			}
		}

		/** 将元素放入有尺寸限制的集合中 */
		public static void putObjInDicWithMax<T>(long key,T obj,LongObjectMap<T> dic,int max,Comparison<T> compare)
		{
			if(max<=0)
			{
				dic.put(key,obj);
			}
			else
			{
				if(dic.contains(key))
				{
					dic.put(key,obj);
				}
				else
				{
					if(dic.size()>=max)
					{
						long leastKey=findLeastOfDic(dic,compare);

						dic.remove(leastKey);
					}

					dic.put(key,obj);
				}
			}
		}
	}
}